downstairs/src/region.rs

```diff
@@ -2317,23 +2344,41 @@ impl Region {
             crucible_bail!(ModifyingReadOnlyRegion);
         }
 
-        // Set the extent we will stop flushing at.
         cdt::os__flush__start!(|| job_id);
-        let extent_count = match extent_limit {
+
+        // Drain any dirty extents we're going to flush, while respecting the
+        // extent_limit if one was provided.
+        let dirty_extents: Vec<usize> = match extent_limit {
             Some(el) => {
                 if el > self.def.extent_count().try_into().unwrap() {
                     crucible_bail!(InvalidExtent);
                 }
-                el + 1 // +1 because our loop is exclusive.
+
+                // TODO someday we can use drain_filter here when it finally
+                // gets out of nightly...
+                let dirty_extents = self
+                    .dirty_extents
+                    .iter()
+                    .copied()
+                    .filter(|x| *x <= el)
+                    .collect();
+
+                for eid in &dirty_extents {
+                    self.dirty_extents.remove(eid);
+                }
```

Comment:
The Downstairs will re-execute jobs until they succeed if errors are seen, so these should only be removed after all potentially erroneous operations have been done, else a rerun of this will not flush anything.

Reply:
ah- i thought downstairs failed out if errors were seen and that a repair would then be needed. i wrote this under that assumption.